internal/users/idgenerator.go

```diff
@@ -0,0 +1,262 @@
+package users
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"slices"
+	"sort"
+
+	"github.com/ubuntu/authd/internal/users/localentries"
+	"github.com/ubuntu/authd/log"
+)
+
+// IDGeneratorIface is the interface that must be implemented by the ID generator.
+type IDGeneratorIface interface {
+	GenerateUID(ctx context.Context, owner IDOwner) (uid uint32, cleanup func(), err error)
+	GenerateGID(ctx context.Context, owner IDOwner) (gid uint32, cleanup func(), err error)
+}
+
+// IDOwner is the interface that must be implemented by the IDs owner to provide
+// the currently used UIDs and GIDs.
+type IDOwner interface {
+	UsedUIDs() ([]uint32, error)
+	UsedGIDs() ([]uint32, error)
+}
+
+// IDGenerator is an ID generator that generates UIDs and GIDs in a specific range.
+type IDGenerator struct {
+	UIDMin uint32
+	UIDMax uint32
+	GIDMin uint32
+	GIDMax uint32
+
+	// IDs generated but not saved to the database yet.
+	// This is used to avoid generating the same ID multiple times.
+	// We don't differentiate between UIDs and GIDs here, because:
+	// * When picking a UID, we avoid IDs which are already used as GIDs,
+	//   because the UID is also used as the GID of the user private group.
+	// * When picking a GID, we avoid IDs which are already used as UIDs,
+	//   because those are also GIDs of the user private groups.
+	pendingIDs []uint32
+}
```

Comment:
> Also, given we use this struct in a locked environment we can avoid any
    further locking, so dropping the mutex. The race tests would tell us if
    in future this will be needed

I assume that with "locked environment" you're referring to the `UserDBLock`. However, IIUC, your implementation of the `UserDBLock` does not prevent multiple goroutines to run in parallel, because of the recursive locking stuff, right?

Reply:
`UserDBLock` well no, it's the manager itself now... Technically it's correct to keep the mutex since we're in two different classes, so I was unsure about, but since we moved this inside the users package, I would consider it more a private helper implementation.

On that regards, I'm still undecided whether to embed the generator or not inside the manager though, since I like the fact that the data that it manages is in a different struct and so are the private methods, but at the same time it uses the manager data... So... Don't know. For testing purposes though we could just reuse all the logic we have, a part the fact it does not go through the users data. But I feel that's work for a next iteration IMHO.